src/main/java/fi/jubic/quanta/models/QuantaAuthenticator.java

```diff
@@ -0,0 +1,73 @@
+package fi.jubic.quanta.models;
+
+import fi.jubic.quanta.dao.ExternalClientDao;
+import fi.jubic.snoozy.auth.implementation.StatefulAuthenticator;
+import fi.jubic.snoozy.auth.implementation.Token;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.stream.Collectors;
+
+@Singleton
+public class QuantaAuthenticator extends StatefulAuthenticator<User> {
+    private List<Token<User>> tokens;
+    private final ExternalClientDao externalClientDao;
+
+    @Inject
+    public QuantaAuthenticator(
+            ExternalClientDao externalClientDao
+    ) {
+        this.tokens = new CopyOnWriteArrayList<>();
+        this.externalClientDao = externalClientDao;
+        reloadExternalClients();
+    }
+
+    @Override
+    public Optional<User> authenticate(String token) {
+        if (super.authenticate(token).isPresent()) {
```

Comment:
`Optional::or` would be a good fit here. This approach ends up with 2 calls to the super class method whenever a regular user token is authenticated.

```java
return super.authenticate(token)
        .or(() -> tokens.stream()
                .filter(t -> t.getToken().equals(token))
                .findFirst()
                .map(Token::getUser)
        )
```

Reply:
Changed the API languages to 1.9 in order to use `.or()`

Reply:
Changed to Java 11